[JSC] Introduce wyhash to StringImpl

hyjorc1 opened the pull request:
#### ff19fc587702c99f818714c97f079c6ae60498bc
<pre>
[JSC] Introduce wyhash to StringImpl
<a href="https://bugs.webkit.org/show_bug.cgi?id=260147">https://bugs.webkit.org/show_bug.cgi?id=260147</a>
rdar://113860312

Reviewed by Yusuke Suzuki.

StringImpl::hashSlowCase() is one of the hottest functions when running
JetStream2. However, the existing hash function is slow for large
string since it&apos;s based on incremental hashing of each character.
This patch introduces wyhash[1] (hashing data in bulks) to StringImpl
with OSS approval (ID OSS-16203), since wyhash provides the best hash
score without quality issues according to SMhasher[2].

In this patch:
1. Rename previous StringHasher.h to SuperFastHash.h.
2. Use SuperFastHash for small strings and wyhash for large strings
in order to gain best performances on JetStream2.

[1] <a href="https://github.com/wangyi-fudan/wyhash">https://github.com/wangyi-fudan/wyhash</a>
[2] <a href="https://github.com/rurban/smhasher">https://github.com/rurban/smhasher</a>

* Source/JavaScriptCore/create_hash_table:
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp.in:
* Source/JavaScriptCore/runtime/JSCell.cpp:
(JSC::reportZappedCellAndCrash):
* Source/JavaScriptCore/runtime/TemplateObjectDescriptor.h:
(JSC::TemplateObjectDescriptor::calculateHash):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
* Source/JavaScriptCore/tools/VMInspector.cpp:
(JSC::VMInspector::dumpSubspaceHashes):
* Source/JavaScriptCore/yarr/hasher.py:
(stringHash):
(finalizeAndMaskTop8Bits):
(superFastHash):
(wyhash):
(wyhash.add64):
(wyhash.multi64):
(wyhash.wymum):
(wyhash.wymix):
(wyhash.convert32BitTo64Bit):
(wyhash.convert16BitTo32Bit):
(wyhash.c2i):
(wyhash.wyr8):
(wyhash.wyr4):
(wyhash.wyr2):
(ceilingToPowerOf2):
(createHashTable):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/Hasher.h:
* Source/WTF/wtf/text/CString.cpp:
(WTF::CString::hash const):
* Source/WTF/wtf/text/StringHash.h:
(WTF::ASCIICaseInsensitiveHash::hash):
* Source/WTF/wtf/text/StringHasher.h:
(WTF::StringHasher::StringHasher):
(WTF::StringHasher::finalize):
(WTF::StringHasher::finalizeAndMaskTop8Bits):
(WTF::StringHasher::addCharactersAssumingAligned): Deleted.
(WTF::StringHasher::addCharacter): Deleted.
(WTF::StringHasher::addCharacters): Deleted.
(WTF::StringHasher::hashWithTop8BitsMasked const): Deleted.
(WTF::StringHasher::hash const): Deleted.
(WTF::StringHasher::computeHashAndMaskTop8Bits): Deleted.
(WTF::StringHasher::computeHash): Deleted.
(WTF::StringHasher::computeLiteralHash): Deleted.
(WTF::StringHasher::computeLiteralHashAndMaskTop8Bits): Deleted.
(WTF::StringHasher::calculateWithRemainingLastCharacter): Deleted.
(WTF::StringHasher::calculateWithTwoCharacters): Deleted.
(WTF::StringHasher::computeHashImpl): Deleted.
(WTF::StringHasher::processPendingCharacter const): Deleted.
* Source/WTF/wtf/text/StringHasherInlines.h: Added.
(WTF::StringHasher::computeHashAndMaskTop8Bits):
(WTF::StringHasher::computeLiteralHashAndMaskTop8Bits):
(WTF::StringHasher::addCharacter):
(WTF::StringHasher::hashWithTop8BitsMasked):
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/SuperFastHash.h: Copied from Source/WTF/wtf/text/StringHasher.h.
(WTF::SuperFastHash::addCharactersAssumingAligned):
(WTF::SuperFastHash::addCharacter):
(WTF::SuperFastHash::addCharacters):
(WTF::SuperFastHash::hashWithTop8BitsMasked const):
(WTF::SuperFastHash::hash const):
(WTF::SuperFastHash::computeHashAndMaskTop8Bits):
(WTF::SuperFastHash::computeHash):
(WTF::SuperFastHash::computeLiteralHash):
(WTF::SuperFastHash::computeLiteralHashAndMaskTop8Bits):
(WTF::SuperFastHash::calculateWithRemainingLastCharacter):
(WTF::SuperFastHash::calculateWithTwoCharacters):
(WTF::SuperFastHash::computeHashImpl):
(WTF::SuperFastHash::processPendingCharacter const):
* Source/WTF/wtf/text/WYHash.h: Added.
(WTF::WYHash::computeHashAndMaskTop8Bits):
(WTF::WYHash::wyrot):
(WTF::WYHash::wymum):
(WTF::WYHash::wymix):
(WTF::WYHash::Reader16Bit::hasDefaultConverter):
(WTF::WYHash::Reader16Bit::convert):
(WTF::WYHash::Reader16Bit::wyr3):
(WTF::WYHash::Reader16Bit::wyr4WithConvert):
(WTF::WYHash::Reader16Bit::wyr4):
(WTF::WYHash::Reader16Bit::wyr8WithConvert):
(WTF::WYHash::Reader16Bit::wyr8):
(WTF::WYHash::Reader8Bit::hasDefaultConverter):
(WTF::WYHash::Reader8Bit::convert):
(WTF::WYHash::Reader8Bit::wyr3):
(WTF::WYHash::Reader8Bit::wyr4WithConvert):
(WTF::WYHash::Reader8Bit::wyr4):
(WTF::WYHash::Reader8Bit::wyr8WithConvert):
(WTF::WYHash::Reader8Bit::wyr8):
(WTF::WYHash::initSeed):
(WTF::WYHash::consume24Characters):
(WTF::WYHash::handleEndCase):
(WTF::WYHash::handleGreaterThan8CharactersCase):
(WTF::WYHash::hash):
(WTF::WYHash::computeHashImpl):
* Source/WTF/wtf/unicode/UTF8Conversion.cpp:
(WTF::Unicode::calculateStringHashAndLengthFromUTF8MaskingTop8Bits):
* Source/WebCore/bindings/scripts/Hasher.pm:
(finalizeAndMaskTop8Bits):
(superFastHash):
(uint64_add):
(uint64_multi):
(GenerateHashValue): Deleted.
* Source/WebCore/contentextensions/DFAMinimizer.cpp:
* Source/WebCore/contentextensions/HashableActionList.h:
(WebCore::ContentExtensions::HashableActionList::HashableActionList):
* Source/WebCore/platform/SharedStringHash.cpp:
(WebCore::computeSharedStringHashInline):
* Source/WebCore/platform/graphics/WidthCache.h:
(WebCore::WidthCache::SmallStringKey::SmallStringKey):
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/StringHasher.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WTF/SuperFastHash.cpp: Copied from Tools/TestWebKitAPI/Tests/WTF/StringHasher.cpp.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WTF/WYHash.cpp: Added.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/266929@main">https://commits.webkit.org/266929@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a497c148b001ee1870b55aaf70f44aa8020804a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15129 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15434 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15787 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16884 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14220 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15268 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17949 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15532 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16850 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15311 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15768 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12863 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17612 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13048 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13649 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20614 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/12962 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14128 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13817 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17064 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14383 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14379 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12184 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/15324 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13658 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3905 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17997 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/15557 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1833 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14219 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3715 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ *
+ */
```
Algorithm and implementation is completely different, so we should,

1. Have 2-clause BSD with only Apple copyright,
2. and https://github.com/wangyi-fudan/wyhash/blob/master/LICENSE

hyjorc1 replied:
I am not sure I was doing it correctly. Need to double check.

Constellation commented on Source/WTF/wtf/unicode/UTF8Conversion.cpp:
```diff
@@ -129,7 +131,7 @@ bool convertUTF8ToUTF16ReplacingInvalidSequences(const char* source, const char*
 unsigned calculateStringHashAndLengthFromUTF8MaskingTop8Bits(const char* data, const char* dataEnd, unsigned& dataLength, unsigned& utf16Length)
 {
-    StringHasher stringHasher;
+    WTF::Vector<UChar> buffer;
```
We should make it better.

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+    static unsigned computeHashImpl(const T* characters, unsigned characterCount)
+    {
+        ASSERT(sizeof(decltype(Converter::convert(std::declval<T>()))) == 2);
+        Vector<uint16_t> buffer;
```
That's too costly. Let's not do it.

hyjorc1 commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+        for (size_t i = 0; i < characterCount; ++i)
+            buffer.append(Converter::convert(*characters++));
+        return hash<uint16_t, Reader16Bit<uint16_t>>(buffer.data(), characterCount);
+    }
```
If the caller request to pass a Converter instead of the default one, we would take this path which might be costly.

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+    static constexpr unsigned computeHashImpl(const T* characters, unsigned characterCount)
+    {
+        if constexpr (sizeof(T) == 2)
+            return hash<T, Reader16Bit<T>>(characters, characterCount);
```
Let's use
```
if constexpr (sizeof(T) == 2)
    return hash<T, Reader16Bit<T>>(characters, characterCount);
else
    return hash<T, Reader8Bit<T>>(characters, characterCount);
```
otherwise, else side does not become `constexpr`.

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+            if (std::is_constant_evaluated() || enable_is_constant_evaluated)
+                return static_cast<uint32_t>(p[0]) | (static_cast<uint32_t>(p[1]) << 16);
+            uint32_t v;
+            memcpy(&v, p, 2);
```
Let's use `UnalignedAccess.h` and `uint32_t v = unalignedLoad<uint16_t>(p);`

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+                    | (static_cast<uint64_t>(p[3]) << 48);
+            }
+            uint64_t v;
+            memcpy(&v, p, 4);
```
Let's use `UnalignedAccess.h` and `uint64_t v = unalignedLoad<uint32_t>(p);`

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+            }
+            uint64_t v;
+            memcpy(&v, p, 8);
+            return v;
```
Let's use `return unalignedLoad<uint64_t>(p);`

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+            }
+            uint32_t v;
+            memcpy(&v, p, 4);
+            return v;
```
Let's use `return unalignedLoad<uint32_t>(p);`

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+            ASSERT(sizeof(T) == 2);
+            uint64_t firstByte = static_cast<uint8_t>(p[0]);
+            uint64_t secondByte = static_cast<uint8_t>(p[0] >> 8);
+            return (firstByte << 16) | (secondByte << 8) | secondByte;
```
This looks strange, why?
I think it should be
```
return static_cast<uint64_t>(static_cast<uint16_t>(p[0]));
```
Also, we need to have `CPU(BIG_ENDIAN)` side implementation (see wyhash's implementation).

Constellation replied:
Discussed offline. This is not `wyr2`, it is `wyr3`. So, we should rename this function to avoid confusion.

hyjorc1 replied:
Here `wyr2` is equivalent to `wyr3` in https://github.com/wangyi-fudan/wyhash/blob/77e50f267fbc7b8e2d09f2d455219adb70ad4749/wyhash.h#L116C3-L116C3, which is used for both `BIG_ENDIAN` and `LITTLE_ENDIAN`.  So, just want to be consistent with original algorithm. I will rename `wry2` to `wyr3`.

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+            ASSERT(sizeof(T) == 2);
+            if (std::is_constant_evaluated() || enable_is_constant_evaluated) {
+                return static_cast<uint32_t>(p[0])
+                    | (static_cast<uint32_t>(p[1]) << 16);
```
Why not using `uint64_t` for expanding?
```
return static_cast<uint64_t>(static_cast<uint16_t>(p[0])) | (static_cast<uint64_t>(static_cast<uint16_t>(p[1])) << 16);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+        r *= *B;
+        *A = static_cast<uint64_t>(r);
+        *B = static_cast<uint64_t>(r >> 64);
+#else
```
Let's remove else side and always use `UInt128`. (Including `Int128.h`).
And because this is not C

```
UInt128 r = static_cast<UInt128>(*A) * *B;
*A = static_cast<uint64_t>(r);
*B = static_cast<uint64_t>(r >> 64);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+        ALWAYS_INLINE static constexpr uint64_t wyr2(const T* p)
+        {
+            ASSERT(sizeof(T) == 1);
+            return static_cast<uint64_t>(p[0]) << 16;
```
This looks incorrect. This case should be,

```
return return static_cast<uint64_t>(static_cast<uint8_t>(p[0]));
```

Constellation replied:
Discussed offline. This function is not `wyr2`, it is `wyr3`.

Constellation replied:
So, make it
```
return static_cast<uint64_t>(static_cast<uint8_t>(p[0])) << 16;
```

hyjorc1 replied:
Ditto, will rename it to `wyr3`.

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+                return static_cast<uint64_t>(p[0])
+                    | (static_cast<uint64_t>(p[1]) << 16)
+                    | (static_cast<uint64_t>(p[2]) << 32)
+                    | (static_cast<uint64_t>(p[3]) << 48);
```
Wrap them with `static_cast<uint16_t>()` since some characters can be signed (so, let's avoid sign extension explicitly).
```
                return static_cast<uint64_t>(static_cast<uint16_t>(p[0]))
                    | (static_cast<uint64_t>(static_cast<uint16_t>(p[1])) << 16)
                    | (static_cast<uint64_t>(static_cast<uint16_t>(p[2])) << 32)
                    | (static_cast<uint64_t>(static_cast<uint16_t>(p[3])) << 48);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+
+        ALWAYS_INLINE static constexpr uint64_t wyr8(const T* p)
+        {
+            ASSERT(sizeof(T) == 2);
```
Let's make it
```
static_assert(sizeof(T) == 2);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+
+        ALWAYS_INLINE static constexpr uint64_t wyr4(const T* p)
+        {
+            ASSERT(sizeof(T) == 2);
```
Let's make it
```
static_assert(sizeof(T) == 2);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+    struct Reader16Bit {
+        ALWAYS_INLINE static constexpr uint64_t wyr2(const T* p)
+        {
+            ASSERT(sizeof(T) == 2);
```
Let's make it
```
static_assert(sizeof(T) == 2);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+    struct Reader8Bit {
+        ALWAYS_INLINE static constexpr uint64_t wyr2(const T* p)
+        {
+            ASSERT(sizeof(T) == 1);
```
Let's make it
```
static_assert(sizeof(T) == 1);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+
+        ALWAYS_INLINE static constexpr uint64_t wyr4(const T* p)
+        {
+            ASSERT(sizeof(T) == 1);
```
Let's make it
```
static_assert(sizeof(T) == 1);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+
+        ALWAYS_INLINE static constexpr uint64_t wyr8(const T* p)
+        {
+            ASSERT(sizeof(T) == 1);
```
Let's make it
```
static_assert(sizeof(T) == 1);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+        {
+            ASSERT(sizeof(T) == 1);
+            if (std::is_constant_evaluated() || enable_is_constant_evaluated)
+                return static_cast<uint32_t>(p[0]) | (static_cast<uint32_t>(p[1]) << 16);
```
Let's wrap it with `static_cast<uint8_t>()` to avoid accidental sign extension.
```
return static_cast<uint32_t>(static_cast<uint8_t>(p[0])) | (static_cast<uint32_t>(static_cast<uint8_t>(p[1])) << 16);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+                return static_cast<uint64_t>(p[0])
+                    | (static_cast<uint64_t>(p[1]) << 16)
+                    | (static_cast<uint64_t>(p[2]) << 32)
+                    | (static_cast<uint64_t>(p[3]) << 48);
```
Let's wrap each of item with `static_cast<uint8_t>()`.
```
                return static_cast<uint64_t>(static_cast<uint8_t>(p[0]))
                    | (static_cast<uint64_t>(static_cast<uint8_t>(p[1])) << 16)
                    | (static_cast<uint64_t>(static_cast<uint8_t>(p[2])) << 32)
                    | (static_cast<uint64_t>(static_cast<uint8_t>(p[3])) << 48);
```

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+        auto wyr4 = Reader::wyr4;
+        auto wyr8 = Reader::wyr8;
+
+        const unsigned byteCount = characterCount << 1;
```
Let's use `uint64_t` for byteCount to avoid accidental overflow.

Constellation replied:
Do it, `const uint64_t byteCount = static_cast<uint64_t>(characterCount) << 1;`.

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+                const unsigned offset = (((byteCount >> 3) << 2)) >> 1;
+                a = (wyr4(p) << 32) | wyr4(p + offset);
+                p += characterCount - 2;
+                b = (wyr4(p) << 32) | wyr4(p - offset);
```
Is this offset computation correct? If `p` is `const UChar*`, then using byteCount-based offset sounds not correct.

hyjorc1 replied:
Yes, this is correct.

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+
+        if (LIKELY(byteCount <= 16)) {
+            if (LIKELY(byteCount >= 4)) {
+                const unsigned offset = (((byteCount >> 3) << 2)) >> 1;
```
This is `>> 1`, but it is only true when `T` is `UChar` (this is not correct when `T` is `LChar`).

Constellation replied:
Ignore this, this is correct.

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+                } while (LIKELY(i > 48));
+                seed ^= see1 ^ see2;
+            }
+            while (LIKELY(i > 16)) {
```
Isn't it `UNLIKELY`?

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+            }
+        } else {
+            size_t i = byteCount;
+            if ((i > 48)) {
```
Isn't it `UNLIKELY`?

Constellation commented on Source/WTF/wtf/text/WYHasher.h:
```diff
@@ -0,0 +1,264 @@
+public:
+    static constexpr const unsigned flagCount = 8;
+    static constexpr const unsigned maskHash = (1U << (sizeof(unsigned) * 8 - flagCount)) - 1;
+    static constexpr const uint64_t secret[4] = { 0xa0761d6478bd642full, 0xe7037ed1a0b428dbull, 0x8ebc6af09c88c6e3ull, 0x589965cc75374cc3ull };
```
static constexpr is enough (`const` is not necessary).

Constellation commented on Source/WTF/wtf/text/WYHash.h:
```diff
@@ -0,0 +1,275 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
```
Add wyhash LICENSE text here https://github.com/wangyi-fudan/wyhash/blob/master/LICENSE

hyjorc1 commented on Source/JavaScriptCore/yarr/hasher.py:
```diff
@@ -41,31 +74,189 @@ def stringHash(str):
+#     'dontEnum',
+#     'dontDelete',
+#     'dontDeleteDontEnum',
+# ], "staticDontDeleteDontEnumTableIndex"))
```
Will remove these after we choose the threshold for small string.

Constellation commented on Source/WTF/wtf/text/StringHasher.h:
```diff
@@ -263,71 +106,14 @@ class StringHasher {
-    UChar m_pendingCharacter { 0 };
-    bool m_hasPendingCharacter { false };
+    unsigned bufferSize { 0 };
+    FixedVector<UChar> buffer;
```
Use `std::array<UChar, smallStringThreshold>` instead.

Constellation replied:
And you should follow to `m_` naming.

Constellation commented on Source/WTF/wtf/text/StringHasher.h:
```diff
@@ -263,71 +106,14 @@ class StringHasher {
-    unsigned m_hash { stringHashingStartValue };
-    UChar m_pendingCharacter { 0 };
-    bool m_hasPendingCharacter { false };
+    unsigned bufferSize { 0 };
```
Ditto, you should use m_ naming.

hyjorc1 commented on Source/JavaScriptCore/yarr/hasher.py:
```diff
@@ -41,31 +74,189 @@ def stringHash(str):
+    return v
+
+
+def createHashTable(keys, hashTableName, compactSize=0):
```
This is used to create the `CompactHashIndex`s in `JSDollarVM.cpp`. Remove this or add a comment here?

Constellation commented on Source/WTF/wtf/text/StringHasher.h:
```diff
@@ -263,71 +106,14 @@ class StringHasher {
+    unsigned numberOfProcessedCharacters { 0 };
+    uint64_t seed { 0 };
+    uint64_t see1 { 0 };
+    uint64_t see2 { 0 };
```
Ditto, you should follow to `m_` naming.

Constellation commented on Source/WTF/wtf/text/WYHash.h:
```diff
@@ -0,0 +1,275 @@
+
+namespace WTF {
+
+#define FORCE_CONVERT_IN_READ 0
```
Let's use `static constexpr bool` instead.

Constellation commented on Source/WTF/wtf/text/WYHash.h:
```diff
@@ -0,0 +1,275 @@
+            static_assert(sizeof(T) == 2);
+            if (std::is_constant_evaluated() || FORCE_CONVERT_IN_READ)
+                return wyr4WithConvert(p);
+            return unalignedLoad<uint32_t>(p);
```
Does it work for BigEndian?

hyjorc1 replied:
No, fixed.

Constellation commented on Source/WTF/wtf/text/WYHash.h:
```diff
@@ -0,0 +1,275 @@
+            uint16_t v = convert(p[0]);
+            uint64_t firstByte = static_cast<uint8_t>(v);
+            uint64_t secondByte = static_cast<uint8_t>(v >> 8);
+            return (firstByte << 16) | (secondByte << 8) | secondByte;
```
Does it work for BigEndian?

hyjorc1 replied:
Yes, for `wyr3`, the original implementation uses this for both `BIG_ENDAN` and `LITTLE_ENDAN`.
https://github.com/wangyi-fudan/wyhash/blob/77e50f267fbc7b8e2d09f2d455219adb70ad4749/wyhash.h#L116

hyjorc1 replied:
btw, `BIG_ENDAN` or `LITTLE_ENDAN` only affects `memcpy`.

Constellation commented on Source/WTF/wtf/text/WYHash.h:
```diff
@@ -0,0 +1,275 @@
+            static_assert(sizeof(T) == 2);
+            if (std::is_constant_evaluated() || FORCE_CONVERT_IN_READ)
+                return wyr8WithConvert(p);
+            return unalignedLoad<uint64_t>(p);
```
Does it work for BigEndian?

hyjorc1 replied:
No, fixed.

Constellation commented on Source/WTF/wtf/text/StringHasherInlines.h:
```diff
@@ -0,0 +1,117 @@
+
+#pragma once
+
+#include <unicode/umachine.h>
```
you do not need to include this.

Constellation commented on Source/JavaScriptCore/create_hash_table:
```diff
@@ -194,60 +197,209 @@ sub calcCompactHashSize()
+    my $string = $_[0];
+    my @chars = split(/ */, $string);
+    my $charCount = scalar @chars;
+    if ($charCount <= 48) {
```
Is it correct?

Constellation commented on Source/JavaScriptCore/tools/JSDollarVM.cpp:
```diff
@@ -743,15 +743,11 @@ JSC_DEFINE_CUSTOM_SETTER(testStaticAccessorPutter, (JSGlobalObject* globalObject
 
-static const struct CompactHashIndex staticCustomAccessorTableIndex[9] = {
-    { -1, -1 },
+static const struct CompactHashIndex staticCustomAccessorTableIndex[5] = {
+    { 0, 4 },
```
Are they correctly updated with 72 threshold?

hyjorc1 replied:
Yes, just re-computed and it's correct.

Constellation commented on Source/JavaScriptCore/tools/JSDollarVM.cpp:
```diff
@@ -762,7 +758,7 @@ static const struct HashTableValue staticCustomAccessorTableValues[3] = {
 
 static const struct HashTable staticCustomAccessorTable =
-    { 3, 7, PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum | PropertyAttribute::CustomAccessor, nullptr, staticCustomAccessorTableValues, staticCustomAccessorTableIndex };
+    { 3, 3, PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum | PropertyAttribute::CustomAccessor, nullptr, staticCustomAccessorTableValues, staticCustomAccessorTableIndex };
```
Ditto

hyjorc1 replied:
Yes, just re-computed and it's correct.

Constellation commented on Source/JavaScriptCore/tools/JSDollarVM.cpp:
```diff
@@ -843,15 +839,12 @@ JSC_DEFINE_CUSTOM_SETTER(testStaticValuePutterSetFlag, (JSGlobalObject* globalOb
-    { -1, -1 },
-    { -1, -1 },
-    { 3, -1 },
+static const struct CompactHashIndex staticCustomValueTableIndex[5] = {
```
Ditto.

hyjorc1 replied:
Yes, just re-computed and it's correct.

Constellation commented on Source/JavaScriptCore/tools/JSDollarVM.cpp:
```diff
@@ -862,7 +855,7 @@ static const struct HashTableValue staticCustomValueTableValues[4] = {
 
 static const struct HashTable staticCustomValueTable =
-    { 4, 7, PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum | PropertyAttribute::CustomValue, nullptr, staticCustomValueTableValues, staticCustomValueTableIndex };
+    { 4, 3, PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum | PropertyAttribute::CustomValue, nullptr, staticCustomValueTableValues, staticCustomValueTableIndex };
```
Ditto

hyjorc1 replied:
Yes, just re-computed and it's correct.

Constellation commented on Source/JavaScriptCore/tools/JSDollarVM.cpp:
```diff
@@ -940,15 +933,12 @@ JSC_DEFINE_HOST_FUNCTION(staticDontDeleteDontEnumMethod, (JSGlobalObject*, CallF
 }
 
-static const struct CompactHashIndex staticDontDeleteDontEnumTableIndex[8] = {
+static const struct CompactHashIndex staticDontDeleteDontEnumTableIndex[5] = {
```
Ditto

hyjorc1 replied:
Yes, just re-computed and it's correct.

Constellation commented on Source/JavaScriptCore/tools/JSDollarVM.cpp:
```diff
@@ -958,7 +948,7 @@ static const struct HashTableValue staticDontDeleteDontEnumTableValues[3] = {
 
 static const struct HashTable staticDontDeleteDontEnumTable =
-    { 3, 7, PropertyAttribute::DontEnum | PropertyAttribute::DontDelete, nullptr, staticDontDeleteDontEnumTableValues, staticDontDeleteDontEnumTableIndex };
+    { 3, 3, PropertyAttribute::DontEnum | PropertyAttribute::DontDelete, nullptr, staticDontDeleteDontEnumTableValues, staticDontDeleteDontEnumTableIndex };
```
Ditto

hyjorc1 replied:
Yes, just re-computed and it's correct.

Constellation commented on Source/WTF/wtf/text/WYHash.h:
```diff
@@ -0,0 +1,333 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
```
Let's separate these two comments. Like,
```
 */
/* 
 * License header from wyhash https://github.com/wangyi-fudan/wyhash/blob/master/LICENSE
 * 
 * This is free and unencumbered software released into the public domain.
```

Constellation reviewed (approved):
r=me with one thing.

1. Let's mention to OSS-16203 ID in commit message (wyhash OSS approval)

hyjorc1 committed:
[JSC] Introduce wyhash to StringImpl
https://bugs.webkit.org/show_bug.cgi?id=260147
rdar://113860312

Reviewed by Yusuke Suzuki.

StringImpl::hashSlowCase() is one of the hottest functions when running
JetStream2. However, the existing hash function is slow for large
string since it's based on incremental hashing of each character.
This patch introduces wyhash[1] (hashing data in bulks) to StringImpl
with OSS approval (ID OSS-16203), since wyhash provides the best hash
score without quality issues according to SMhasher[2].

In this patch:
1. Rename previous StringHasher.h to SuperFastHash.h.
2. Use SuperFastHash for small strings and wyhash for large strings
in order to gain best performances on JetStream2.

[1] https://github.com/wangyi-fudan/wyhash
[2] https://github.com/rurban/smhasher

* Source/JavaScriptCore/create_hash_table:
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp.in:
* Source/JavaScriptCore/runtime/JSCell.cpp:
(JSC::reportZappedCellAndCrash):
* Source/JavaScriptCore/runtime/TemplateObjectDescriptor.h:
(JSC::TemplateObjectDescriptor::calculateHash):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
* Source/JavaScriptCore/tools/VMInspector.cpp:
(JSC::VMInspector::dumpSubspaceHashes):
* Source/JavaScriptCore/yarr/hasher.py:
(stringHash):
(finalizeAndMaskTop8Bits):
(superFastHash):
(wyhash):
(wyhash.add64):
(wyhash.multi64):
(wyhash.wymum):
(wyhash.wymix):
(wyhash.convert32BitTo64Bit):
(wyhash.convert16BitTo32Bit):
(wyhash.c2i):
(wyhash.wyr8):
(wyhash.wyr4):
(wyhash.wyr2):
(ceilingToPowerOf2):
(createHashTable):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/Hasher.h:
* Source/WTF/wtf/text/CString.cpp:
(WTF::CString::hash const):
* Source/WTF/wtf/text/StringHash.h:
(WTF::ASCIICaseInsensitiveHash::hash):
* Source/WTF/wtf/text/StringHasher.h:
(WTF::StringHasher::StringHasher):
(WTF::StringHasher::finalize):
(WTF::StringHasher::finalizeAndMaskTop8Bits):
(WTF::StringHasher::addCharactersAssumingAligned): Deleted.
(WTF::StringHasher::addCharacter): Deleted.
(WTF::StringHasher::addCharacters): Deleted.
(WTF::StringHasher::hashWithTop8BitsMasked const): Deleted.
(WTF::StringHasher::hash const): Deleted.
(WTF::StringHasher::computeHashAndMaskTop8Bits): Deleted.
(WTF::StringHasher::computeHash): Deleted.
(WTF::StringHasher::computeLiteralHash): Deleted.
(WTF::StringHasher::computeLiteralHashAndMaskTop8Bits): Deleted.
(WTF::StringHasher::calculateWithRemainingLastCharacter): Deleted.
(WTF::StringHasher::calculateWithTwoCharacters): Deleted.
(WTF::StringHasher::computeHashImpl): Deleted.
(WTF::StringHasher::processPendingCharacter const): Deleted.
* Source/WTF/wtf/text/StringHasherInlines.h: Added.
(WTF::StringHasher::computeHashAndMaskTop8Bits):
(WTF::StringHasher::computeLiteralHashAndMaskTop8Bits):
(WTF::StringHasher::addCharacter):
(WTF::StringHasher::hashWithTop8BitsMasked):
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/SuperFastHash.h: Copied from Source/WTF/wtf/text/StringHasher.h.
(WTF::SuperFastHash::addCharactersAssumingAligned):
(WTF::SuperFastHash::addCharacter):
(WTF::SuperFastHash::addCharacters):
(WTF::SuperFastHash::hashWithTop8BitsMasked const):
(WTF::SuperFastHash::hash const):
(WTF::SuperFastHash::computeHashAndMaskTop8Bits):
(WTF::SuperFastHash::computeHash):
(WTF::SuperFastHash::computeLiteralHash):
(WTF::SuperFastHash::computeLiteralHashAndMaskTop8Bits):
(WTF::SuperFastHash::calculateWithRemainingLastCharacter):
(WTF::SuperFastHash::calculateWithTwoCharacters):
(WTF::SuperFastHash::computeHashImpl):
(WTF::SuperFastHash::processPendingCharacter const):
* Source/WTF/wtf/text/WYHash.h: Added.
(WTF::WYHash::computeHashAndMaskTop8Bits):
(WTF::WYHash::wyrot):
(WTF::WYHash::wymum):
(WTF::WYHash::wymix):
(WTF::WYHash::Reader16Bit::hasDefaultConverter):
(WTF::WYHash::Reader16Bit::convert):
(WTF::WYHash::Reader16Bit::wyr3):
(WTF::WYHash::Reader16Bit::wyr4WithConvert):
(WTF::WYHash::Reader16Bit::wyr4):
(WTF::WYHash::Reader16Bit::wyr8WithConvert):
(WTF::WYHash::Reader16Bit::wyr8):
(WTF::WYHash::Reader8Bit::hasDefaultConverter):
(WTF::WYHash::Reader8Bit::convert):
(WTF::WYHash::Reader8Bit::wyr3):
(WTF::WYHash::Reader8Bit::wyr4WithConvert):
(WTF::WYHash::Reader8Bit::wyr4):
(WTF::WYHash::Reader8Bit::wyr8WithConvert):
(WTF::WYHash::Reader8Bit::wyr8):
(WTF::WYHash::initSeed):
(WTF::WYHash::consume24Characters):
(WTF::WYHash::handleEndCase):
(WTF::WYHash::handleGreaterThan8CharactersCase):
(WTF::WYHash::hash):
(WTF::WYHash::computeHashImpl):
* Source/WTF/wtf/unicode/UTF8Conversion.cpp:
(WTF::Unicode::calculateStringHashAndLengthFromUTF8MaskingTop8Bits):
* Source/WebCore/bindings/scripts/Hasher.pm:
(finalizeAndMaskTop8Bits):
(superFastHash):
(uint64_add):
(uint64_multi):
(GenerateHashValue): Deleted.
* Source/WebCore/contentextensions/DFAMinimizer.cpp:
* Source/WebCore/contentextensions/HashableActionList.h:
(WebCore::ContentExtensions::HashableActionList::HashableActionList):
* Source/WebCore/platform/SharedStringHash.cpp:
(WebCore::computeSharedStringHashInline):
* Source/WebCore/platform/graphics/WidthCache.h:
(WebCore::WidthCache::SmallStringKey::SmallStringKey):
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/StringHasher.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WTF/SuperFastHash.cpp: Copied from Tools/TestWebKitAPI/Tests/WTF/StringHasher.cpp.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WTF/WYHash.cpp: Added.
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/266929@main
